SPEC.md

```diff
@@ -0,0 +1,334 @@
+```
+
+## Endpoint Descriptions
+
+### Start Operation
+
+Start an arbirary length operation.
+The response of the operation may be delivered synchronously (inline), or asynchronously, via a provided callback or the
+[Get Operation Result](#get-operation-result) endpoint.
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}`
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}`
+
+#### Method
+
+`POST`
+
+#### Query Parameters
+
+- `callback_url`: Optional. If the operation is asynchronous and the handler supports callbacks, it should invoke this URL once the operation's result is available.
+
+#### Request Body
+
+The body may contain arbitrary data. Headers should specify content type and encoding.
+
+#### Response Codes
+
+- `200 OK`: Operation completed successfully.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.
+
+- `204 No Content`: Operation completed successfully with an empty result.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Empty.
+
+- `482 Operation Failed`: Operation completed as `failed` or `canceled`.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**: A JSON serialized [`Failure`](#failure) object.
+
+- `201 Created`: Operation was started an will completed asynchronously.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+
+  **Body**: A JSON serialized [`OperationStarted`](#operationstarted) object.
+
+### Cancel Operation
+
+Request to cancel an operation.
+The operation may later complete as canceled or any other outcome.
+Handlers should ignore multiple cancelations of the same operation and return successfully if cancelation was already
+requested.
+
+#### Method
+
+`POST`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/cancel`
+
+#### Response Codes
+
+- `202 Accepted`: Cancelation request accepted.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Result
+
+Retrieve operation result.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/result`
+
+#### Query Parameters
+
+- `wait_deadline`: Optional. ISO 8601 date indicating the waiting period for a result.
+  If not specified, and the operation is still running, the request should resolve immediately with the current
+  operation status.
+
+#### Response Codes
+
+- `200 OK`: Operation completed successfully.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.
+
+- `204 No Content`: Operation completed successfully with an empty result or is still running.
```

Comment:
This is a special case where I think we should not return 204 even if the handler returned no data. I should be able to == 200 this call and know the body is what was sent even with no body. Whether there's a body or _from the handler_ I don't think affects the protocol/status in this case.

Reply:
I'd be fine with that, I was debating this and ended up with 204 but don't have a strong preference.

Reply:
Yeah, I don't really have a good way to explain why 204 shouldn't be used in _all_ no-content situations. But basically in a situation where the protocol is unaware of the body size, it should just be 200 whether 1 byte body or 0 byte body. 204 is a protocol thing (i.e. "I _know_ it will be no body").

Reply:
I couldn't find anywhere that says you should prefer 204 over 200 for empty responses.
I could definitely change to 200, if we want to also support 204 we can add that later but I wouldn't encourage it for simplicity.